### PR TITLE
Integrated training: Framework updates and Modular DFP pipeline implementation

### LINKED ISSUE #595: [FEA] Add supporting infrastructure and primitives for Morpheus training pipelines.
The current roadmap for implementing training in Morpheus has two phases:

1. MRC framework updates to support reusable, configurable, modules.
2. Morpheus implementations of modularized application frameworks, eg. Tao, Mlflow, etc... each acting as a configurable black box.

The end result will support a variety of new functionality:

1. Control messages which can be used for data loading, training, and inference:
- Indicate how the data should be acquired: raw, file based, or external source and have it retrieved via loader
- Initiate multiple training tasks, within a single persistent pipeline, each with its own configuration parameters
- Run inference on a specified collection of data
2. Support for running inference or training pipelines, or a combination of the two
3. Support for online/real time mirroring of inference data to an online training pipeline
4. Support for pre-defined or customizable stages for drift detection within a data stream, which can trigger downstream tasks (such as model building).

```[tasklist]
### Proof of concept translating existing DFP pipeline into a composite mrc-module (completed by bsuryadevara)
- [x] Extraction of work units from Morpheus stages into stored modules
- [x] Extend existing example to fully encompass inference and training DFP pipelines as a single module. 
- [x] Creation of baseline utilities for loading an arbitrary module as a single Morpheus stages
- [x] Examples, documentation, and unit testing for DFP PoC code
```

```[tasklist]
### Added support for 'ControlMessage' Objects
- [x] Define API and Data Type
- [x] c++ implementation + python bindings
```

```[tasklist]
### Creation of Human-In-The-Loop modules
- [x] Data labeling: provide a simple (probably beta/PoC) module to illustrate the process of allowing real time labeling of data which can be fed back into an online training pipeline.
- [x] Model training kick off: provide a simple (probably beta/PoC) module that provides an interface to issue training tasks to a running pipeline.
```

```[tasklist]
### Creation of Automatic / Hands Free training modules
- [x] Data loader capable processing ControlMesages - general idea is to have the incoming data element describe the data resource the loader should prepare.
- [x] Support loader implementations that can be registered and loaded by the DataLoader class
- [ ] Thresholding / Trigger component - think drift detection, will allow for automatic triggering of downstream tasks when some criteria is met.
- [x] Updated training components for DFP to support training run's triggered off of ControlMessages
- [x] MLflow integration for tracking iterative or HPO style training in morpheus -- 'always on' modules that can initiate new training tasks when a command tag is received.
```

```[tasklist]
### Examples, documentation, and unit tests for training configurations
- [x] Refactor DFP modular pipeline to use ControlMessages, and support training and inference tasks
- [ ] Refactor Phishing pipeline to use ControlMessages, support training and inference tasks, and provide human in the loop GUI
- [x] Define configuration API based on experience implementing PoC use cases
- [ ] Put DFP and Phishing pipelines under functional testing harness to measure correctness and performance
```

```[tasklist]
### Reach items, which may go into 23.06 [nice to have] -- Some items depend on TAO updates
- [ ] Define specific PoC use cases and implement them as Tao training pipelines
- [ ] Import `tao-client` interface into a set of `mrc::modules` to support directly interacting with Tao backend service
- [ ] (human in the loop) Model sign off -- provide a simple module that provides an interface for a human to approve a given model so that it can be rolled into active service in an inference pipeline.
- [ ] Automated End-to-end testing of Tao training pipelines
- [ ] Isolation testing for training pipelines to ensure high availability/recoverability
- [ ] Define a configuration schema for composing a Tao AutoEncoder training pipeline in Morpheus
- [ ] Feature implementation/interface definitions: data source creation, model training kick off, finished model validation, finished model deployment, 'models as data' (?)
```

```[tasklist]
### Benchmarking
- [x] Understand and minimize performance impact of mirroring training data away from production data streams
```

- [X] I agree to follow this project's Code of Conduct
- [X] I have searched the [open feature requests](https://github.com/nv-morpheus/Morpheus/issues?q=is%3Aopen+is%3Aissue+label%3A%22feature+request%22%2Cimprovement%2Cenhancement) and have found no duplicates for this feature request

### PR DESCRIPTION
**Pull request by drobison00**
Resolves #595

### CODE REVIEW

File: morpheus/_lib/src/io/loaders/file_list.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                files.push_back(filename);
+            }
+        }
+    }
+
+    // TODO(Devin): Improve robustness
+    // For now, a directory listing will just create an updated control message with the new file list
+    // nlohmann::json task_properties = {{"properties", {"loader_id", "file", "files", files}}};
+
+    // TODO(Devin): This is a placeholder implementation -- need to add file list as a payload
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/io/loaders/file_list.cpp:62]: Remove?

File: morpheus/_lib/src/python_modules/messages.cpp [hunk 2]
```diff
@@ -314,10 +332,50 @@ PYBIND11_MODULE(messages, m)
...
         .def_property_readonly("probs", &MultiResponseProbsMessageInterfaceProxy::probs);
 
-#ifdef VERSION_INFO
-    m.attr("__version__") = MACRO_STRINGIFY(VERSION_INFO);
-#else
-    m.attr("__version__") = "dev";
-#endif
+    py::enum_<ControlMessageType>(_module, "ControlMessageType")
+        .value("INFERENCE", ControlMessageType::INFERENCE)
+        .value("NONE", ControlMessageType::INFERENCE)
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/python_modules/messages.cpp:337]: `NONE` should be first here

File: morpheus/_lib/include/morpheus/stages/preallocate.hpp [hunk 3]
```diff
@@ -71,12 +94,38 @@ struct PreallocateStageInterfaceProxy
         std::vector<std::tuple<std::string, TypeId>> needed_columns);
 };
 
-// Explicit instantiations
-template class PreallocateStage<MessageMeta>;
-template class PreallocateStage<MultiMessage>;
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/stages/preallocate.hpp:76]: Why did you remove the explicit instantiation here? Can we use this on more types?

File: morpheus/_lib/tests/io/test_io.hpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+#pragma once
+
+#include "../test_morpheus.hpp"  // IWYU pragma: associated
+
+namespace morpheus::test {
+
+TEST_CLASS(DataLoaderRegistry);
+TEST_CLASS(Loader);
```
Comment (mdemoret-nv) [inline morpheus/_lib/tests/io/test_io.hpp:25]: Is this file needed?

File: morpheus/_lib/tests/modules/test_data_loader_module.cpp [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+// interactions.
+// TEST_F(TestDataLoaderModule, EndToEndFileDataLoaderTest)
+//{
+//    using namespace mrc::modules;
+//    using namespace mrc;
+//
+//    using sp_msg_ctrl_t = std::shared_ptr<MessageControl>;
+//
+//    std::size_t packet_count{0};
+//
```
Comment (mdemoret-nv) [inline morpheus/_lib/tests/modules/test_data_loader_module.cpp:51]: Large comment

File: morpheus/_lib/tests/test_morpheus.cpp [hunk 6]
```diff
@@ -17,19 +17,51 @@
...
+namespace {
+std::string accum_merge(std::string lhs, std::string rhs)
+{
+    if (lhs.empty())
+    {
+        return std::move(rhs);
+    }
+
+    return std::move(lhs) + "," + std::move(rhs);
+}
```
Comment (mdemoret-nv) [inline morpheus/_lib/tests/test_morpheus.cpp:46]: We have a similar function in `StringUtil`. Can that be used?

File: morpheus/_lib/__init__.py [hunk 7]
```diff
@@ -14,3 +14,4 @@
 from . import common
 from . import messages
 from . import stages
+from . import modules
```
Comment (mdemoret-nv) [inline morpheus/_lib/__init__.py:17]: Keep sorted

File: morpheus/loaders/__init__.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Morpheus loader definitions, each loader is automatically registered when imported and will be available for use.
+"""
+
+from morpheus.loaders import file_to_df_loader
+from morpheus.loaders import fsspec_loader
+
+__all__ = []
```
Comment (mdemoret-nv) [inline morpheus/loaders/__init__.py:21]: This needs to contain strings with the names `file_to_df_loader`, and `fsspec_loader`

File: morpheus/messages/message_control.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    """
+    MessageControl is an object that serves as a specification of the tasks to be executed in a pipeline workflow.
+    The MessageControl is passed between stages of the pipeline, with each stage executing the tasks specified in
+    the MessageControl configuration.
+
+    MessageControl is capable of carrying payload of the MessageMeta type.
+    """
+
+    def __init__(self, *arg, **kwargs):
+        super().__init__(*arg, **kwargs)
```
Comment (mdemoret-nv) [inline morpheus/messages/message_control.py:29]: If there is no python implementation, this just needs to import the C++ class:
```suggestion
from morpheus._lib.messages import MessageControl
```

File: morpheus/modules/file_batcher.py [hunk 10]
```diff
@@ -83,50 +147,97 @@ def on_data(file_objects: fsspec.core.OpenFiles):
...
         full_names = []
-        file_objs = []
-        for (ts, file_object) in ts_and_files:
+        for (ts, file_name) in ts_and_files:
             timestamps.append(ts)
-            full_names.append(file_object.full_name)
-            file_objs.append(file_object)
+            full_names.append(file_name)
 
+        # df = pd.DataFrame()
```
Comment (mdemoret-nv) [inline morpheus/modules/file_batcher.py:161]: Excess comment

File: morpheus/modules/filter_control_message.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+            # if the data_type don't match the filter_task_type.
+            if (filter_data_type and filter_data_type != filter):
+                return None
+
+        return control_message
+
+    def node_fn(obs: mrc.Observable, sub: mrc.Subscriber):
+        obs.pipe(ops.map(on_data), ops.filter(lambda x: x is not None)).subscribe(sub)
+
+    node = builder.make_node(FILTER_CONTROL_MESSAGE, mrc.core.operators.build(node_fn))
```
Comment (mdemoret-nv) [inline morpheus/modules/filter_control_message.py:86]: You can replace this with: 
```suggestion
    node = builder.make_node(FILTER_CONTROL_MESSAGE, ops.map(on_data), ops.filter(lambda x: x is not None))
```

File: morpheus/stages/input/control_message_kafka_source_stage.py [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+from morpheus.pipeline.preallocator_mixin import PreallocatorMixin
+from morpheus.pipeline.single_output_source import SingleOutputSource
+from morpheus.pipeline.stream_pair import StreamPair
+from morpheus.stages.input.kafka_source_stage import AutoOffsetReset
+
+logger = logging.getLogger(__name__)
+
+
+@register_stage("from-cm-kafka", modes=[PipelineModes.AE])
+class ControlMessageKafkaSourceStage(PreallocatorMixin, SingleOutputSource):
```
Comment (mdemoret-nv) [inline morpheus/stages/input/control_message_kafka_source_stage.py:37]: It would be much better to reimplement this or `KafkaSourceStage` to reuse the same class so code is not duplicated

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
Completed the review of the changes in `morpheus` folder. Continuing on to the `examples` section next. In general, the following comments applied to multiple locations:

- Replace uses of `#pragma GCC visibility push(default)` with `MORPHEUS_EXPORT` on the necessary classes and methods
- Remove the logging from all of the destructors. This will just pollute the VLOG
- Replace all uses of `boost::filesystem` with `std::filesystem`
- Cleanup all `TODO` comments
- When using `__all__` in an `__init__.py` file, you need to add all of the imported names into the `__all__` list. See `morpheus/messages/__init__.py`
- This PR has changes from #712 which still has many outstanding comments.

File: morpheus/_lib/include/morpheus/messages/control.hpp [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+
+  private:
+    static const std::string s_config_schema;  // NOLINT
+
+    ControlMessageType m_cm_type{ControlMessageType::NONE};
+    std::shared_ptr<MessageMeta> m_payload{nullptr};
+    std::map<std::string, ControlMessageType> m_task_type_map{{"inference", ControlMessageType::INFERENCE},
+                                                              {"training", ControlMessageType::TRAINING}};
+
+    nlohmann::json m_tasks{};
```
[the remark this answers is not in the document]
Comment (drobison00) [reply morpheus/_lib/include/morpheus/messages/control.hpp:130]: Currently, no. I'd thought about having a priority, or some type of sorting metric, but I didn't have any required use case, and it could be added later.

File: morpheus/_lib/include/morpheus/messages/control.hpp [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+     */
+    bool has_metadata(const std::string& key) const;
+
+    // TODO(Devin): Allow for a default value
+    /**
+     * @brief Get the metadata value for a given key
+     * @param key
+     * @return
+     */
+    const nlohmann::json get_metadata(const std::string& key) const;
```
[the remark this answers is not in the document]
Comment (drobison00) [reply morpheus/_lib/include/morpheus/messages/control.hpp:95]: Yes, when you get back the json object, you can call get<type>() . In the contexts where its returned to python land its directly converted to a dictionary.

File: morpheus/_lib/include/morpheus/io/data_loader.hpp [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    virtual std::shared_ptr<MessageControl> load(std::shared_ptr<MessageControl> message, nlohmann::json task);
+
+  protected:
+    nlohmann::json config() const;
+
+  private:
+    nlohmann::json m_config{};
+};
+
+class DataLoader
```
[the remark this answers is not in the document]
Comment (drobison00) [reply morpheus/_lib/include/morpheus/io/data_loader.hpp:52]: DataLoader is a higher level component that is effectively a collection of loaders; individual loaders are for a specific task, like loading a file. Will add additional comments.

File: morpheus/_lib/src/io/loaders/file.cpp [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+#include <boost/filesystem.hpp>
+#include <glog/logging.h>
+#include <nlohmann/json.hpp>
+#include <pybind11/pybind11.h>
+#include <pymrc/utilities/object_cache.hpp>
+
+#include <fstream>
+#include <memory>
+
+namespace {}
```
[the remark this answers is not in the document]
Comment (drobison00) [reply morpheus/_lib/src/io/loaders/file.cpp:32]: Removed.

File: morpheus/_lib/src/io/loaders/file.cpp [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+        }
+        std::transform(extension.begin(), extension.end(), extension.begin(), ::tolower);
+
+        VLOG(5) << "Loading file: " << file.dump(2);
+
+        // TODO(Devin): Any extensions missing?
+        auto current_df = mod_cudf.attr("DataFrame")();
+        if (extension == "csv")
+        {
+            current_df = mod_cudf.attr("read_csv")(path.string());
```
[the remark this answers is not in the document]
Comment (drobison00) [reply morpheus/_lib/src/io/loaders/file.cpp:82]: Individual loaders get their associated task, so you can attach somewhat arbitrary data as part of the load task, and parse it in the loader implementation.

File: morpheus/_lib/src/io/loaders/file_list.cpp (hunk 1 above)
Comment (drobison00) [reply morpheus/_lib/src/io/loaders/file_list.cpp:62]: Actually going to remove this in favor of the fsspec loader

File: morpheus/_lib/include/morpheus/stages/preallocate.hpp (hunk 3 above)
Comment (drobison00) [reply morpheus/_lib/include/morpheus/stages/preallocate.hpp:76]: I can probably roll this back; at the time I started updating things there was a bug here that was causing linking errors; I think David has since fixed it.

File: morpheus/_lib/tests/io/test_io.hpp (hunk 4 above)
Comment (drobison00) [reply morpheus/_lib/tests/io/test_io.hpp:25]: Yes, I think so. It fits in the existing format, and helps to create segregated training blocks in the test output.

File: morpheus/_lib/tests/test_morpheus.cpp (hunk 6 above)
Comment (drobison00) [reply morpheus/_lib/tests/test_morpheus.cpp:46]: Yeah, that function works.

File: morpheus/loaders/__init__.py (hunk 8 above)
Comment (drobison00) [reply morpheus/loaders/__init__.py:21]: I don't think so. Neither of those classes is meant to be used directly. Just importing them here causes them to be registered with the loader registry.

File: morpheus/stages/input/control_message_kafka_source_stage.py (hunk 12 above)
Comment (drobison00) [reply morpheus/stages/input/control_message_kafka_source_stage.py:37]: @bsuryadevara Can you refresh me on the details here; I know we'd talked about this at some length, but there was a reason that it didn't make sense. Or am I miss-remembering?

Comment (bsuryadevara) [reply morpheus/stages/input/control_message_kafka_source_stage.py:37]: There are multiple reasons why we needed to write different implementations:

Current DFP modules implementation pipeline is running with the use_cpp mode set to true. When we use the existing KafkaSourceStage, it ends up using the c++ implementation. This implementation uses `cudf.io.json` which is split the control message into multiple columns.

We attempted to use use_cpp=False mode, but unfortunately, we don't have the option with the existing Python implementation to choose the engine type or pass extra arguments.

KafkaSourceStage returns MessageMeta with a batch of messages. However, in this case, the ControlMessage  itself is a metadata that instructs how and where to load the data, as well as what tasks need to be performed. In this scenario, we need to restrict set the pipeline batch size to 1 by all means if above works.

Comment (drobison00) [reply morpheus/stages/input/control_message_kafka_source_stage.py:37]: @bsuryadevara Perfect, thanks.

**Review by mdemoret-nv — APPROVED** [review]
Approving to not be a blocker to getting this in. Once all comments are fixed, merge and we can push the release.
